src/lib/nvme_pt_ioctl.c

```diff
@@ -355,19 +358,70 @@ static int opal_level0_disc_pt(struct sed_device *device)
 
 void opal_deinit_pt(struct sed_device *dev)
 {
+	struct opal_device *opal_dev = dev->priv;
+
 	if (dev->fd != 0) {
 		close(dev->fd);
 		dev->fd = 0;
 	}
 
-	if (dev->priv != NULL) {
-		free(dev->priv);
-		dev->priv = NULL;
+	if (opal_dev != NULL) {
+		if (opal_dev->req_buf != NULL) {
+			free(opal_dev->req_buf);
+			opal_dev->req_buf = NULL;
+		}
+		if (opal_dev->resp_buf != NULL) {
+			free(opal_dev->resp_buf);
+			opal_dev->resp_buf = NULL;
+		}
+		free(opal_dev);
+		opal_dev = NULL;
 	}
 
 	opal_parser_deinit();
 }
 
+static uint64_t tper_prop_to_val(struct sed_device *dev, const char *tper_prop_name)
+{
+	struct sed_tper_properties *tper = &dev->discv.sed_tper_props;
+
+	for (int j = 0; j < NUM_TPER_PROPS; j++) {
+		if (strncmp(tper_prop_name, tper->property[j].key_name,
+				strlen(tper_prop_name)) == 0)
+			return tper->property[j].value;
+	}
+
+	SEDCLI_DEBUG_MSG("Invalid TPer property name!\n");
+
+	return -EINVAL;
+}
+
+static int resize_io_buf(struct opal_device *dev, uint64_t size)
+{
+       if (dev == NULL)
+               return -EINVAL;
+
+       if (dev->req_buf != NULL)
+               free(dev->req_buf);
+       if (dev->resp_buf != NULL)
+               free(dev->resp_buf);
+
+       dev->req_buf = malloc(sizeof(uint8_t) * size);
+       if (dev->req_buf == NULL)
+               return -ENOMEM;
+       dev->resp_buf = malloc(sizeof(uint8_t) * size);
+       if (dev->resp_buf == NULL)
+               return -ENOMEM;
```

Comment:
when code follows this path then req_buf will store pointer to allocated memory., while resp_buf will not. I think this is error prone, because it leaves deallocation to the caller. 
To make things simplier I would suggest to do only one allocation of 2*size. this will simplify function: for example:

....
/* allocate buffer for req and resp*/
ptr = malloc(sizeof(uint8_t) * size * 2);
if (!ptr) {
....
}
memset (ptr, 0, size);
dev->req_buf = &ptr[0];
dev->resp_buf = &ptr[size];

Reply:
That's true...I should have included a free(req_buf), in-case resp buf memory allocation failed.
But I like the idea you have suggested and will use it.

Reply:
Also, instead of using sizeof(uint8_t), we can use sizeof(*dev->req_buf).